File: ai/linear_model.R
args <- commandArgs(trailingOnly = TRUE)

if (length(args) != 3) {
  stop(
    paste(
      "Usage: Rscript linear_model.R",
      "<filename> <x_column> <y_column>"
    )
  )
}

filename <- args[1]
x_column <- args[2]
y_column <- args[3]

# Read the dataset
dataset <- read.csv(filename)

# Check that the requested columns exist
missing_columns <- setdiff(
  c(x_column, y_column),
  names(dataset)
)

if (length(missing_columns) > 0) {
  stop(
    paste(
      "Missing column(s):",
      paste(missing_columns, collapse = ", ")
    )
  )
}

# Create a simple data frame using x and y
df <- data.frame(
  x = dataset[[x_column]],
  y = dataset[[y_column]]
)

df <- na.omit(df)

# Fit the linear regression model
model <- lm(y ~ x, data = df)

slope <- coef(model)[2]
intercept <- coef(model)[1]
r <- cor(df$x, df$y)

predicted <- predict(model)
mse <- mean((df$y - predicted)^2)

# Print the required statistics
cat("Linear Regression Results\n")
cat("-------------------------\n")
cat("Slope:", round(slope, 4), "\n")
cat("Intercept:", round(intercept, 4), "\n")
cat("Pearson correlation coefficient (r):", round(r, 4), "\n")
cat("Mean Squared Error (MSE):", round(mse, 4), "\n")

library(ggplot2)

annotation_text <- paste0(
  "y = ",
  round(slope, 2),
  "x + ",
  round(intercept, 2),
  "\nr = ",
  round(r, 2),
  "\nMSE = ",
  round(mse, 2)
)

regression_plot <- ggplot(df, aes(x = x, y = y)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE) +
  annotate(
    "text",
    x = min(df$x),
    y = max(df$y),
    label = annotation_text,
    hjust = 0,
    vjust = 1
  ) +
  labs(
    title = paste(y_column, "vs.", x_column),
    x = x_column,
    y = y_column
  ) +
  theme_minimal()

# Use the exact filename required by the instructor
ggsave(
  "regression_plot_r.png",
  plot = regression_plot,
  width = 8,
  height = 6,
  dpi = 300
)

cat("Plot saved as regression_plot_r.png\n")

File: ai/linear_model.py
import sys

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from scipy.stats import linregress
from sklearn.metrics import mean_squared_error


def main():
    if len(sys.argv) != 4:
        print(
            "Usage: python linear_model.py "
            "<filename> <x_column> <y_column>"
        )
        sys.exit(1)

    filename = sys.argv[1]
    x_column = sys.argv[2]
    y_column = sys.argv[3]

    # Read the dataset
    dataset = pd.read_csv(filename)

    # Check that the requested columns exist
    if x_column not in dataset.columns or y_column not in dataset.columns:
        print("Error: one or both column names were not found.")
        print("Available columns:", list(dataset.columns))
        sys.exit(1)

    # Keep only the selected columns and remove missing values
    analysis_data = dataset[[x_column, y_column]].dropna()

    x = analysis_data[x_column].to_numpy()
    y = analysis_data[y_column].to_numpy()

    # Fit the linear regression model
    slope, intercept, r_value, p_value, std_error = linregress(x, y)

    # Calculate predicted values and MSE
    y_pred = slope * x + intercept
    mse = mean_squared_error(y, y_pred)

    # Print the required statistics
    print("Linear Regression Results")
    print("-------------------------")
    print(f"Slope: {slope:.4f}")
    print(f"Intercept: {intercept:.4f}")
    print(f"Pearson correlation coefficient (r): {r_value:.4f}")
    print(f"Mean Squared Error (MSE): {mse:.4f}")

    # Sort values so the fitted line is drawn correctly
    order = np.argsort(x)

    # Create the enhanced regression plot
    plt.figure(figsize=(8, 6))
    plt.scatter(x, y, label="Observed data")
    plt.plot(x[order], y_pred[order], label="Fitted line")

    annotation = (
        f"y = {slope:.2f}x + {intercept:.2f}\n"
        f"r = {r_value:.2f}\n"
        f"MSE = {mse:.2f}"
    )

    plt.text(
        0.05,
        0.95,
        annotation,
        transform=plt.gca().transAxes,
        verticalalignment="top",
        bbox={"boxstyle": "round", "alpha": 0.8},
    )

    plt.xlabel(x_column)
    plt.ylabel(y_column)
    plt.title(f"{y_column} vs. {x_column}")
    plt.legend()
    plt.tight_layout()

    # Use the exact filename required by the instructor
    plt.savefig("regression_plot_python.png", dpi=300)
    plt.close()

    print("Plot saved as regression_plot_python.png")


if __name__ == "__main__":
    main()